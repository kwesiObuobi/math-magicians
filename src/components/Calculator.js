import React, { useState } from 'react';
import PropTypes from 'prop-types';
import './calculator.css';
import calculate from '../logic/calculate';

const Calculator = () => {
  const [object, setObject] = useState({
    total: 0,
    next: null,
    operation: null,
  });

  const handleClick = (button) => {
    setObject(calculate(object, button));
  };

  const calculatorBoard = [
    ['AC', '+/-', '%', 'รท'],
    ['7', '8', '9', 'x'],
    ['4', '5', '6', '-'],
    ['1', '2', '3', '+'],
    ['0', '.', '='],
  ];

  return (
    <div className="calculator">
      <div className="result">
        <span>0</span>
      </div>
      <div className="row">
        <Key name="AC" />
        <Key name="+/-" />
        <Key name="%" />
        <Key name="รท" tag="yellow" />
      </div>
      <div className="row">
        <Key name="7" />
        <Key name="8" />
        <Key name="9" />
        <Key name="X" tag="yellow" />
      </div>
      <div className="row">
        <Key name="4" />
        <Key name="5" />
        <Key name="6" />
        <Key name="-" tag="yellow" />
      </div>
      <div className="row">
        <Key name="1" />
        <Key name="2" />
        <Key name="3" />
        <Key name="+" tag="yellow" />
      </div>
      <div className="row">
        <Key name="0" tag="zero" />
        <Key name="." />
        <Key name="=" tag="yellow" />
      </div>
    </div>
  );
};

const Key = ({ name, tag }) => <button type="button" className={`key ${tag}`}>{ name }</button>;

Key.defaultProps = {
  tag: '',
};

Key.propTypes = {
  name: PropTypes.string.isRequired,
  tag: PropTypes.string,
};

export default Calculator;
