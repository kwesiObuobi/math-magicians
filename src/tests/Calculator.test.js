import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Calculator from '../components/Calculator';

describe('Calculator Component', () => {
  it('renders the calculator component', () => {
    render(<Calculator />);
    expect(screen.getByText('AC')).toBeInTheDocument();
    expect(screen.getByText('=')).toBeInTheDocument();
    expect(screen.queryByRole('link')).toBeNull();
  });

  it('renders the calculator component correctly', () => {
    render(<Calculator />);
    expect(screen.getByText('AC')).toBeInTheDocument();
    expect(screen.getByText('=')).toBeInTheDocument();
    expect(screen.queryByRole('link')).toBeNull();
  });

  it('should display 2 numbers to be added', () => {
    render(<Calculator />);
    const button1 = screen.getByText('9');
    const plusBtn = screen.getByText('+');
    const button2 = screen.getByText('9');

    fireEvent.click(button1);
    fireEvent.click(plusBtn);
    fireEvent.click(button2);

    const result = screen.getByText('9 + 9');

    expect(result).toBeInTheDocument();
  });

  it('should show the result of two added numbers', () => {
    render(<Calculator />);
    const button1 = screen.getByText('9');
    const plusBtn = screen.getByText('+');
    const button2 = screen.getByText('9');
    const equalsBtn = screen.getByText('=');

    fireEvent.click(button1);
    fireEvent.click(plusBtn);
    fireEvent.click(button2);
    fireEvent.click(equalsBtn);

    const result = screen.getByText('18');

    expect(result).toBeInTheDocument();
  });

  it('should correctly render the Calculator component', () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
