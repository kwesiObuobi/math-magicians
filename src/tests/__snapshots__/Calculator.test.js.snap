// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator Component should correctly render the Calculator component 1`] = `
<div
  className="calculator-container"
>
  <h1
    className="calc-header"
  >
    Let's do some math!
  </h1>
  <div
    className="calculator"
  >
    <div
      className="result"
    >
      <span>
        0
         
         
      </span>
    </div>
    <div
      className="row"
    >
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        AC
      </button>
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        +/-
      </button>
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        %
      </button>
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        รท
      </button>
    </div>
    <div
      className="row"
    >
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        7
      </button>
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        8
      </button>
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        9
      </button>
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        x
      </button>
    </div>
    <div
      className="row"
    >
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        4
      </button>
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        5
      </button>
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        6
      </button>
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        -
      </button>
    </div>
    <div
      className="row"
    >
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        1
      </button>
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        2
      </button>
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        3
      </button>
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        +
      </button>
    </div>
    <div
      className="row"
    >
      <button
        className="key zero"
        onClick={[Function]}
        type="button"
      >
        0
      </button>
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        .
      </button>
      <button
        className="key "
        onClick={[Function]}
        type="button"
      >
        =
      </button>
    </div>
  </div>
</div>
`;
